### PR TITLE
Add analytics page

### PR DESCRIPTION
**Pull request by ruben1**

### CODE REVIEW

**Comment by ruben1** [comment]
I would shoot myself if I had to review this

**Comment by khkyler** [comment]
You have merge conflicts so it looks like you will be reviewing this after all....

File: src/client/app/modules/analytics/analytics.controller.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+          vm.formInfo.methods = res.methods;
+          vm.formInfo.persistence = res.persistence;
+          vm.formInfo.data = res.data;
+          vm.formInfo.methodKeys = Object.keys(vm.formInfo.methods);
+          vm.formInfo.businessLogic = res.businessLogic;
+          vm.tableHead = vm.formInfo.data[0];
+          if(res.analytics) {
+            vm.analytics = res.analytics;
+          }
+          console.log('analytics data from db' , res.analytics);
```
Comment (AndrewSouthpaw) [inline src/client/app/modules/analytics/analytics.controller.js]: do we need this?

File: src/client/app/modules/analytics/analytics.controller.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+    function prepareData() {
+      var position = {
+        'GET': 0,
+        'POST': 1,
+        'PUT': 2,
+        'DELETE': 3
+      };
+
+      var defaultValues = {
```
Comment (AndrewSouthpaw) [inline src/client/app/modules/analytics/analytics.controller.js]: It's unclear (from reading your code only) what these default values mean. I would add some documentation here to explain. Same with `position` above.

File: src/client/app/modules/analytics/analytics.controller.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+      var xAxis = ['days', 'Mon', 'Tue', 'Wed', 'Thur', 'Fri', 'Sat', 'Sun'];
+      data.unshift(xAxis);
+
+      return data;
+    }
+
+    function renderChart(data) {
+      //var data = vm.prepareData();
+      console.log('data from renderChart', data);
```
Comment (AndrewSouthpaw) [inline src/client/app/modules/analytics/analytics.controller.js]: still needed?

File: src/client/app/modules/analytics/analytics.controller.js [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+              position: 'outer-middle'
+            }
+          }
+        } 
+      });
+    }
+
+
+  }
+}).call(this);
```
Comment (AndrewSouthpaw) [inline src/client/app/modules/analytics/analytics.controller.js]: provide newline at EOF

File: src/client/app/modules/analytics/analytics.jade [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  thead
+    tr
+      td(ng-repeat="(key, value) in vm.tableHead")
+        a(href='#', ng-click="vm.sortType = key; vm.sortReverse = !vm.sortReverse")
+          | {{ key }}
+          span.fa.fa-caret-down(ng-show="vm.sortType == '{{ key }}' && !vm.sortReverse")
+          span.fa.fa-caret-up(ng-show="vm.sortType == '{{ key }}' && vm.sortReverse")
+  tbody
+    tr(ng-repeat='user in vm.formInfo.data | orderBy:vm.sortType:vm.sortReverse | filter:vm.searchData')
+      td(ng-repeat='column in user') {{ column }}
```
Comment (AndrewSouthpaw) [inline src/client/app/modules/analytics/analytics.jade]: newline at EOF

File: src/client/app/modules/analytics/analytics.jade [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+      .input-group-addon
+        i.fa.fa-search
+      input.form-control(type='text', placeholder='Search', ng-model='vm.searchData')
+table.table.table-bordered.table-striped
+  thead
+    tr
+      td(ng-repeat="(key, value) in vm.tableHead")
+        a(href='#', ng-click="vm.sortType = key; vm.sortReverse = !vm.sortReverse")
+          | {{ key }}
+          span.fa.fa-caret-down(ng-show="vm.sortType == '{{ key }}' && !vm.sortReverse")
```
Comment (AndrewSouthpaw) [inline src/client/app/modules/analytics/analytics.jade]: can we use non-coercive comparator (`===`) here?

File: src/client/app/modules/analytics/analytics.jade [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+          span.fa.fa-caret-up(ng-show="vm.sortType == '{{ key }}' && vm.sortReverse")
```
Comment (AndrewSouthpaw) [inline src/client/app/modules/analytics/analytics.jade]: as above

File: src/client/app/modules/analytics/chart.directive.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+      }
+    };
+
+  }
+}).call(this);
+
+//refactor returned object
+
+//link
+  //insert logic to render chart
```
Comment (AndrewSouthpaw) [inline src/client/app/modules/analytics/chart.directive.js]: newline

File: src/client/app/modules/analytics/chart.directive.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+    return {
+      templateUrl: '/html/modules/analytics/chart.directive.html',
+      restrict: 'E',
+      scope: true,
+      transclude: true,
+      controller: 'AnalyticsController',
+      controllerAs: 'vm',
+      link: function(scope, element, attrs) {
+      }
```
Comment (AndrewSouthpaw) [inline src/client/app/modules/analytics/chart.directive.js]: if link isn't used, let's delete it.

File: src/client/app/modules/analytics/analytics.controller.js [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    function prepareData() {
+      var position = {
+        'GET': 0,
+        'POST': 1,
+        'PUT': 2,
+        'DELETE': 3
+      };
+
+      var defaultValues = {
+        '1': 0,
```
Comment (AndrewSouthpaw) [inline src/client/app/modules/analytics/analytics.controller.js]: is there a reason we can't have the days be zero-indexed?

File: src/server/api/endpointRouter/endpointRouterController.js [hunk 11]
```diff
@@ -9,59 +9,80 @@ var getData = function(req, res, next) {
...
-          for (var i=0; i<endpoint.data.length; i++) {
-            var dataPoint = endpoint.data[i];
-            if (dataPoint.id === queryID) {
-              return res.status(200).json(dataPoint);
+      if (!endpoint.analytics) {
+        endpoint.analytics = {};
+      }
+      if(!endpoint.analytics[method]) {
+        endpoint.analytics[method] = {};
+      }
```
Comment (AndrewSouthpaw) [inline src/server/api/endpointRouter/endpointRouterController.js]: would it make more sense to always initialize a new endpoint with an `analytics` property with all the methods set on it?

File: src/server/api/endpointRouter/endpointRouterController.js [hunk 12]
```diff
@@ -194,46 +254,67 @@ var deleteData = function(req, res, next) {
...
+                console.log('passes conditional');
+                //delete datapoint passed by the user
+                Endpoint.update({ 'username': username, 'route': route }, {$pull: {'data': deleteQuery}}, updateHandler);
+              }
             }
+            return res.status(500).end();
           }
-          return res.status(500).end();
+        } else {
+          if(!endpoint.methods[method]) return res.status(500).end();
```
Comment (AndrewSouthpaw) [inline src/server/api/endpointRouter/endpointRouterController.js]: space after `if`

File: src/server/api/endpointRouter/endpointRouterController.js [hunk 13]
```diff
@@ -194,46 +254,67 @@ var deleteData = function(req, res, next) {
...
+          } else {
+            console.log('gets here');
+            var queryID = parseInt(req.query.id);
+            var deleteQuery = {id: queryID};
+            for(var i = 0; i < endpoint.data.length; i++) {
+              var dataPoint = endpoint.data[i];
+              if(dataPoint.id === queryID) {
+                console.log('passes conditional');
+                //delete datapoint passed by the user
+                Endpoint.update({ 'username': username, 'route': route }, {$pull: {'data': deleteQuery}}, updateHandler);
```
Comment (AndrewSouthpaw) [inline src/server/api/endpointRouter/endpointRouterController.js]: space after `{`: `{ $pull ... }}`

File: src/server/api/endpointRouter/endpointRouterController.js [hunk 14]
```diff
@@ -194,46 +254,67 @@ var deleteData = function(req, res, next) {
...
+          if (!req.query.id) {
+            //we need some data to know what to look for
+            return res.status(500).end();
+          } else {
+            console.log('gets here');
+            var queryID = parseInt(req.query.id);
+            var deleteQuery = {id: queryID};
+            for(var i = 0; i < endpoint.data.length; i++) {
+              var dataPoint = endpoint.data[i];
+              if(dataPoint.id === queryID) {
```
Comment (AndrewSouthpaw) [inline src/server/api/endpointRouter/endpointRouterController.js]: space before paren, also on line 300

File: src/server/api/endpointRouter/endpointRouterController.js [hunk 15]
```diff
@@ -194,46 +254,67 @@ var deleteData = function(req, res, next) {
...
-          console.log('gets here');
-          var queryID = parseInt(req.query.id);
-          var deleteQuery = {id: queryID};
-          for(var i = 0; i < endpoint.data.length; i++) {
-            var dataPoint = endpoint.data[i];
-            if(dataPoint.id === queryID) {
-              console.log('passes conditional');
-              //delete datapoint passed by the user
-              Endpoint.update({ 'username': username, 'route': route }, {$pull: {'data': deleteQuery}}, updateHandler);
+      Endpoint.update({ 'username': username, 'route': route }, {$set: {'analytics': endpoint.analytics }}, function(err, numAffected, rawResponse) {
```
Comment (AndrewSouthpaw) [inline src/server/api/endpointRouter/endpointRouterController.js]: adjust inline objects to have buffer spaces

**Comment by AndrewSouthpaw** [comment]
review complete

**Comment by AndrewSouthpaw** [comment]
LGTM.

**Pull request merged**